### PR TITLE
ZEPPELIN-262 Use spark-submit to run spark interpreter process

### PR DESCRIPTION
**Pull request by Leemoonsoo**
https://issues.apache.org/jira/browse/ZEPPELIN-262

This patch make zeppelin uses spark-submit to run spark interpreter process, when SPARK_HOME is defined. This will potentially solve all the configuration problems related to spark interpreter.

#### How to use?

Define SPARK_HOME env variable in conf/zeppelin-env.sh
Then it'll use your SPARK_HOME/bin/spark-submit, so you will not need any additional configuration :-)

#### Backward compatibility

If You have not defined your SPARK_HOME, you still able to run spark interpreter in old (current) way.
However it is not encouraged anymore.

### CODE REVIEW

**Comment by Leemoonsoo** [comment]
Ready to merge. Please review the changes

File: conf/zeppelin-env.sh.template [hunk 1]
```diff
@@ -33,14 +33,8 @@
...
-# export ZEPPELIN_SPARK_CONCURRENTSQL   # Execute multiple SQL concurrently if set true. false by default.
-# export ZEPPELIN_SPARK_MAXRESULT       # Max number of SparkSQL result to display. 1000 by default.
-
-# Options read in YARN client mode
-# export HADOOP_CONF_DIR         		# yarn-site.xml is located in configuration directory in HADOOP_CONF_DIR.
-
-# Pyspark (supported with Spark 1.2.1 and above)
-# To configure pyspark, you need to set spark distribution's path to 'spark.home' property in Interpreter setting screen in Zeppelin GUI
-# export PYSPARK_PYTHON          		# path to the python command. must be the same path on the driver(Zeppelin) and all workers.
-# export PYTHONPATH              		# extra PYTHONPATH.
```
Comment (felixcheung) [inline conf/zeppelin-env.sh.template]: Is there a reason the last half is taken out?

Comment (Leemoonsoo) [inline conf/zeppelin-env.sh.template]: brought back those lines.

**Comment by bzz** [comment]
This is awesome improvement, thank you @Leemoonsoo
Looks great to me.

**Comment by Leemoonsoo** [comment]
I have pushed more commits, that handles pyspark. Please review them, too.

File: bin/interpreter.sh [hunk 2]
```diff
@@ -72,77 +72,60 @@ fi
...
-      export HADOOP_CONF_DIR="/etc/hadoop/conf"
+    export PYTHONPATH="$SPARK_HOME/python/:$PYTHONPATH"
+    export PYTHONPATH="$SPARK_HOME/python/lib/py4j-0.8.2.1-src.zip:$PYTHONPATH"    
+  else
+    # add Hadoop jars into classpath
+    if [[ -n "${HADOOP_HOME}" ]]; then
+      # Apache
+      addEachJarInDir "${HADOOP_HOME}/share"
+
+      # CDH
```
Comment (vinayshukla) [inline bin/interpreter.sh]: CDH? Is there hadoop distribution specific path?

File: bin/interpreter.sh [hunk 3]
```diff
@@ -72,77 +72,60 @@ fi
...
+    # This will evantually passes SPARK_APP_JAR to classpath of SparkIMain
+    ZEPPELIN_CLASSPATH=${SPARK_APP_JAR}
 
-  # autodetect HADOOP_CONF_HOME by heuristic
-  if [[ -n "${HADOOP_HOME}" ]] && [[ -z "${HADOOP_CONF_DIR}" ]]; then
-    if [[ -d "${HADOOP_HOME}/etc/hadoop" ]]; then
-      export HADOOP_CONF_DIR="${HADOOP_HOME}/etc/hadoop"
-    elif [[ -d "/etc/hadoop/conf" ]]; then
-      export HADOOP_CONF_DIR="/etc/hadoop/conf"
+    export PYTHONPATH="$SPARK_HOME/python/:$PYTHONPATH"
```
Comment (felixcheung) [inline bin/interpreter.sh]: should this check to make sure PYTHONPATH is defined / not empty?

**Comment by randerzander** [comment]
@Leemoonsoo how does z.load work with spark-submit? Seems those dependency jars should be added automatically to spark-submit's --jars argument.

**Comment by Leemoonsoo** [comment]
@randerzander  dependency jars downloaded from z.load() is being loaded after SparkContext is created by calling sc.addJar(). So i think it'll not be affected by this change.

File: bin/interpreter.sh (hunk 2 above)
Comment (Leemoonsoo) [inline bin/interpreter.sh]: It's part of heuristic to search and add hadoop jar files

File: bin/interpreter.sh (hunk 3 above)
Comment (Leemoonsoo) [inline bin/interpreter.sh]: I'd like check environment variable in apache#246

**Comment by felixcheung** [comment]
looks good!

**Comment by Leemoonsoo** [comment]
Merging, if there're no more discussions.

**Comment by smusevic** [comment]
Hello,
  I'm testing out `zeppelin-0.5.6-incubating-bin-all.tgz`.

I might be wrong but it seems to me that this change causes:
```
SPARK_CLASSPATH was detected (set to ':/etc/hbase/conf').
This is deprecated in Spark 1.0+.

Please instead use:
 - ./spark-submit with --driver-class-path to augment the driver classpath
 - spark.executor.extraClassPath to augment the executor classpath

16/03/01 08:11:50 WARN spark.SparkConf: Setting 'spark.executor.extraClassPath' to ':/etc/hbase/conf' as a work-around.
16/03/01 08:11:50 ERROR spark.SparkContext: Error initializing SparkContext.
org.apache.spark.SparkException: Found both spark.driver.extraClassPath and SPARK_CLASSPATH. Use only the former.
        at org.apache.spark.SparkConf$$anonfun$validateSettings$6$$anonfun$apply$8.apply(SparkConf.scala:473)
        at org.apache.spark.SparkConf$$anonfun$validateSettings$6$$anonfun$apply$8.apply(SparkConf.scala:471)
        at scala.collection.immutable.List.foreach(List.scala:318)
        at org.apache.spark.SparkConf$$anonfun$validateSettings$6.apply(SparkConf.scala:471)
        at org.apache.spark.SparkConf$$anonfun$validateSettings$6.apply(SparkConf.scala:459)
        at scala.Option.foreach(Option.scala:236)
        at org.apache.spark.SparkConf.validateSettings(SparkConf.scala:459)
        at org.apache.spark.SparkContext.<init>(SparkContext.scala:391)
        at org.apache.zeppelin.spark.SparkInterpreter.createSparkContext(SparkInterpreter.java:339)
        at org.apache.zeppelin.spark.SparkInterpreter.getSparkContext(SparkInterpreter.java:145)
        at org.apache.zeppelin.spark.SparkInterpreter.open(SparkInterpreter.java:465)
        at org.apache.zeppelin.interpreter.ClassloaderInterpreter.open(ClassloaderInterpreter.java:74)
        at org.apache.zeppelin.interpreter.LazyOpenInterpreter.open(LazyOpenInterpreter.java:68)
        at org.apache.zeppelin.interpreter.LazyOpenInterpreter.interpret(LazyOpenInterpreter.java:92)
        at org.apache.zeppelin.interpreter.remote.RemoteInterpreterServer$InterpretJob.jobRun(RemoteInterpreterServer.java:300)
        at org.apache.zeppelin.scheduler.Job.run(Job.java:169)
        at org.apache.zeppelin.scheduler.FIFOScheduler$1.run(FIFOScheduler.java:134)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
        at java.util.concurrent.FutureTask.run(FutureTask.java:262)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:745)
```
when `conf/zeppelin-env.sh` contains `export SPARK_HOME=...`.
Removing the following text from added line `138` in `bin/interpreter.sh`:
```
--driver-class-path "${ZEPPELIN_CLASSPATH_OVERRIDES}:${CLASSPATH}"
```
the issue is resolved, as suggested by this [email](https://mail-archives.apache.org/mod_mbox/incubator-zeppelin-users/201509.mbox/%3CCADRmTZK5giPRhwjFqaQP_1PL08QWyRMTA=XhDg=z59nwPW3D9Q@mail.gmail.com%3E) but something else happens:
```
| z
<console>:22: error: not found: value z
              z
              ^
```
which sadly blocks me from using `z.load("path/to/jar")` which is what I really need to do.
Please note that I do not have access to change any of the files inside `SPARK_HOME`, including any `conf` files residing therein. 

Is there a workaround for this? Am I doing something wrong?
Thanks in advance!
S.

**Comment by Leemoonsoo** [comment]
@smusevic You look like have `export SPARK_CLASSPATH=/etc/hbase/conf` in conf/zeppelin-env.sh
Could you try `export ZEPPELIN_CLASSPATH=/etc/hbase/conf` instead?

**Comment by smusevic** [comment]
@Leemoonsoo thanks for your reply. I most definitely do not have `export SPARK_CLASSPATH=/etc/hbase/conf` in my `conf/zeppelin-env.sh`, double checked just now.
However, is anyone aware of `SPARK_CLASSPATH` getting initialized during `spark-submit`?
`SPARK_CLASSPATH` does get initialized in the `bit/interpreter.sh`, but investigation has revealed that the line in question does not execute if `SPARK_HOME` is set in `conf/zeppelin-env.sh`.
Anyway, my question now would be: if `SPARK_CLASSPATH` is not set when `spark-submit` is executed then Zeppelin using a `SPARK_HOME` set to some value should work?
Thank you in advance!
Regards,
 S.

**Comment by Leemoonsoo** [comment]
@smusevic 
Right, that should work. Set `SPARK_HOME` is preferred way to configure Zeppelin with Spark.

**Comment by smusevic** [comment]
Thanks, it turned out that `SPARK_CLASSPATH` was set in one of the shell files, bag practice... It works fine now, thanks!

File: bin/interpreter.sh [hunk 4]
```diff
@@ -72,77 +72,60 @@ fi
...
-
-    # CDH
-    addJarInDir "${HADOOP_HOME}"
-    addJarInDir "${HADOOP_HOME}/lib"
-  fi
+  if [[ -n "${SPARK_HOME}" ]]; then
+    export SPARK_SUBMIT="${SPARK_HOME}/bin/spark-submit"
+    SPARK_APP_JAR="$(ls ${ZEPPELIN_HOME}/interpreter/spark/zeppelin-spark*.jar)"
+    # This will evantually passes SPARK_APP_JAR to classpath of SparkIMain
+    ZEPPELIN_CLASSPATH=${SPARK_APP_JAR}
```
Comment (weipuz) [inline bin/interpreter.sh]: Hello, when I set SPARK_HOME to my external Spark, I found the zeppelin-interpreter-sparkxxx.log file are gone. I digged further and found if I change this line 79 in interpreter.sh `ZEPPELIN_CLASSPATH=${SPARK_APP_JAR}` to `ZEPPELIN_CLASSPATH+=${SPARK_APP_JAR}` I get all the Spark interpreter log back.  Is this a bug in the code or I miss understood something? 
Regards,
Weipu

Comment (AhyoungRyu) [inline bin/interpreter.sh]: @weipuz Thanks for digging it ! 
I noticed that issue too. ( I set `SPARK_HOME` and can not get spark log file. )
I also changed  `ZEPPELIN_CLASSPATH=${SPARK_APP_JAR}` to `ZEPPELIN_CLASSPATH+=${SPARK_APP_JAR}` as you said. Then finally I can get my `zeppelin-interpreter-spark-***.log` file. As long as it's not an intended implementation, I think we need to fix this.

Comment (Leemoonsoo) [inline bin/interpreter.sh]: I think it need to be fixed :+1:

Comment (AhyoungRyu) [inline bin/interpreter.sh]: @weipuz @Leemoonsoo I pushed a patch for this issue at #769 : ) 
Thanks again @weipuz for reporting this.

**Pull request closed without merging**